web/docs/language/features.md

```diff
@@ -584,6 +584,107 @@ import { isPrismaError, prismaErrorToHttpError } from '@wasp/utils.js'
   }
 ```
 
+## APIs
+
+In Wasp, the default client-server interaction mechanism is through Operations. However, if you need a specific URL method/path, or a specific response, Operations may not be suitable for you. For these cases, you can use an `api`! Best of all, they should look and feel very familiar.
+
+### API
+
+APIs are used to tie a JS function to an HTTP (method, path) pair. They are distinct from Operations, and have no client-side helpers (like `useQuery`).
+
+To create a Wasp API, you must:
+1. Define the APIs NodeJS implementation
+2. Declare the API in Wasp using the `api` declaration
+
+After completing these two steps, you'll be able to call the API from client code (via our Axios wrapper), or from the outside world.
+
+#### Defining the APIs NodeJS implementation
+An API should be implemented as a NodeJS function that takes three arguments.
+1. `req`:  Express Request object
+2. `res`: Express Response object
+3. `context`: An additional context object **injected into the API by Wasp**. This object contains user session information, as well as information about entities. The examples here won't use the context for simplicity purposes. You can read more about it in the [section about using entities in APIs](#using-entities-in-apis).
+
+Here's an example of a simple API:
+```ts title="src/server/apis.ts"
+import { FooBar } from '@wasp/apis/types'
+
+export const fooBar : FooBar = (req, res, context) => {
+  res.set('Access-Control-Allow-Origin', '*') // Example of modifying headers to override Wasp default CORS middleware.
+  res.json({ msg: `Hello, ${context.user?.username || "stranger"}!` })
+}
+```
+
+#### Declaring an API in Wasp
+After implementing your APIs in NodeJS, all that's left to do before using them is tell Wasp about it!
```

Comment:
We should flip the narrative here (and I think we use a similar sentence when we talk about Queries).

The thing is, we want them to first define their Query/Action/API inside the `.wasp` file, and only then go and define its implementation. If they don't do it in this order, they lose all benefits of type support.

Reply:
So I was thinking of adding this note as a short stopgap until a larger doc refresh:
<img width="848" alt="Screenshot 2023-03-13 at 1 24 06 PM" src="https://user-images.githubusercontent.com/523636/224779267-efde0b42-b221-4044-b098-684135a79899.png">

Reply:
> So we have a bit of a chicken and egg probem when explaining this though, since if we say define it in your .wasp file first, we have to reference some file that does not yet exist. Is that a bit confusing?

It's definitely confusing, that's why I wasn't a fan of having our type support work using this hack, but it was the only way at the moment. And unfortunately I don't know how to word it either.

> Do you mean until they are both defined and the next compile, or indefinitely? I'm not sure how it could be the latter. If it is the former, is it so bad that it is temporarily unavailable until they do both and compile? Maybe a note in between sections is enough for now as a stopgap until a larger rewrite to make sure the TS generation stuff is in all the doc sequences correctly?

Until the `.wasp` file is compiled, yes. Not indefinitely.

 It doesn't seem like a big deal, but I'd argue that it's pretty important.
The main benefits of TypeScript are type checking and autocomplete. If they start implementing the API handler without those benefits, they aren't getting anything out of our TS support (and if they want to use TS, that's probably a deal breaker).

The only way it could work is if they:
1. Define the skeleton of the implementation in the TS file (and export it).
2. Define the declaration inside the `.wasp` file 
3. Go back to the TS file and only then start writing the implementation.

In other words, they'd knowingly have to postpone implementing the entire function when they first create it, and if we're asking them to do that, it's simpler to ask them to just add the query inside the `.wasp` file first.

With all that said, I still wouldn't know how to communicate this, so the note can stay.